### ISSUE TITLE
Question about angular center loss

### ISSUE THREAD

**Issue by Youskrpig** [talreiss/Mean-Shifted-Anomaly-Detection#4]
Hi, I see the angular center loss is expressed as follow:
![image](https://user-images.githubusercontent.com/51428142/127951302-e80f3962-04a4-41ad-954a-62aa599cab9f.png)
I'm confused why it is expressed as such math formual? 
and i see the corresponding code is:
        out_1 = model(img1)
        out_2 = model(img2)
        out_1 = out_1 - center
        out_2 = out_2 - center
        center_loss = ((out_1 ** 2).sum(dim=1).mean() + (out_2 ** 2).sum(dim=1).mean())
which seems like the traidional center loss. Could you please explain?

**Comment by thearkamitra** [comment]
I think there were two losses in the paper right?
The authors did use the angular loss in the next line in the code. I might be missing something though.

**Comment by Youskrpig** [comment]
I misunderstood. Actually it is the angular loss.

**Comment by siyuanseever** [comment]
> Hi, I see the angular center loss is expressed as follow:
> ![image](https://user-images.githubusercontent.com/51428142/127951302-e80f3962-04a4-41ad-954a-62aa599cab9f.png)
> I'm confused why it is expressed as such math formual?
> and i see the corresponding code is:
> out_1 = model(img1)
> out_2 = model(img2)
> out_1 = out_1 - center
> out_2 = out_2 - center
> center_loss = ((out_1 ** 2).sum(dim=1).mean() + (out_2 ** 2).sum(dim=1).mean())
> which seems like the traidional center loss. Could you please explain?

I also think about this, and steal can not understand. I think the right code about the angular center loss should be write like this:
``` python
def angular_loss(out_1, out_2, center):
        out_1 = F.normalize(out_1, dim=-1)
        out_2 = F.normalize(out_2, dim=-1)

        center_loss = -((out_1 * center).sum(dim=1).mean() + (out_2 * center).sum(dim=1).mean())

        return center_loss
```
``` python
out_1 = model(img1)
out_2 = model(img2)

center_loss = angular_loss(out_1, out_2, center)

out_1 = out_1 - center
out_2 = out_2 - center

loss = contrastive_loss(out_1, out_2) + center_loss
```

**Comment by talreiss** [comment]
Hi
Please note that optimizing the Euclidean metric when dealing with unit vectors is proportional to optimizing the angular distance multiplied by a constant factor. Therefore, this optimization process is equivalent.
Hope this clarifies your questions.

**Comment by siyuanseever** [comment]
I think I understand what you said. But when you say 'dealing with unit vectors', the code did not has a 'normalize' function before calculate 'the Euclidean metric'.
``` python
        out_1 = model(img1)
        out_2 = model(img2)
        out_1 = out_1 - center
        out_2 = out_2 - center

        center_loss = ((out_1 ** 2).sum(dim=1).mean() + (out_2 ** 2).sum(dim=1).mean())
```
